graphite/graphite.go

```diff
@@ -76,22 +82,31 @@ func (f *graphitePublisher) Publish(contentType string, content []byte, config m
 	} else {
 		gite, err = graphite.NewGraphite(server, port)
 	}
+	defer gite.Disconnect()
 	if err != nil {
 		logger.Errorf("Error Connecting to graphite at %s:%d. Error: %v", server, port, err)
 		return fmt.Errorf("Error Connecting to graphite at %s:%d. Error: %v", server, port, err)
 	}
 	logger.Debug("Connected to %s:%s successfully", server, port)
-	for _, m := range metrics {
+	giteMetrics := make([]graphite.Metric, len(metrics))
+	for i, m := range metrics {
 		key := m.Namespace().Key()
-		data := fmt.Sprintf("%v", m.Data())
-		logger.Debug("Attempting to send %s:%s", key, data)
-		err = gite.SimpleSend(key, data)
-		if err != nil {
-			logger.Errorf("Unable to send metric %s:%s to %s:%d. Error: %s", key, data, server, port, err)
-			return fmt.Errorf("Unable to send metric %s:%s to %s:%d. Error: %s", key, data, server, port, err)
+		for _, tag := range tagsForPrefix {
+			nextTag, ok := m.Tags()[tag]
+			if ok {
+				key = nextTag + "." + key
+			}
 		}
-		logger.Debug("Sent %s, %s", key, data)
+		data := fmt.Sprintf("%v", m.Data())
+		logger.Debug("Metric ready to send %s:%s", key, data)
+		giteMetrics[i] = graphite.NewMetric(key, data, m.Timestamp().Unix())
+	}
+
+	err = gite.SendMetrics(giteMetrics)
+	if err != nil {
+		return fmt.Errorf("Unable to send metrics. Error: %s", err)
```

Comment:
The only possible error from the Graphite library is this one https://github.com/marpaia/graphite-golang/blob/master/graphite.go#L103 - when it writes the buffer to the socket. It's not linked to the name or the value of a metric. So if it fails for one it fails for all. That's why we don't need more information.

I'll put back the log.